docs/design-documents/hal/0002-pinmap-extension.md

```diff
@@ -0,0 +1,83 @@
+# HAL PinMap design document
+
+# Table of contents
+
+- [Description](#Description)
+- [Motivation](#Motivation)
+- [Requirements](#Requirements)
+- [Design choices](#Design-choices)
+- [Questions](#Questions)
+
+<!-- tocstop -->
+
+# Introduction
+
+Adds board introspection capabilities and tools.  
+These tools will help designing better tests for the HAL.
+
+### Overview and background
+
+At the time being, drivers are only tested on a single "default" peripheral. However, some target feature the same peripheral through different blocks' implementations for example the SPI may be provided on a single MCU by its USART, QSPI and SSP peripherals.
+To ensure that the driver's implementation is valid for all these peripherals we want the CI to run the test set on each peripheral using at least one set of pin determined at run time (pin may eventually picked randomly).
+
+### Requirements and assumptions
+
+1. We want to list all pins for a function on a peripheral.
+   For instance, all pins that can be configured as MOSI for SPI1.
+2. We want to list all functions a pin can provide.
+3. We want to list all pins for a form-factor regardless of the function they can provide.
+4. We want a printable name for each form factor pin and a method to get that name.
```

Comment:
Careful... this could increase ROM usage by a decent chunk.

Reply:
It only takes space if you use it. There is no size increase for most applications - https://github.com/ARMmbed/mbed-os/pull/9449#issuecomment-460817982